src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -120,183 +139,190 @@ public void Dispose()
 
         private partial void DisposeWorker();
 
-        internal bool IsAnalyzerDependencyPath(string fullPath)
+        public void AddDependencyLocation(string originalPath)
         {
             CheckIfDisposed();
 
-            lock (_guard)
+            CompilerPathUtilities.RequireAbsolutePath(originalPath, nameof(originalPath));
+            var simpleName = PathUtilities.GetFileName(originalPath, includeExtension: false);
+            string resolvedPath = originalPath;
+            IAnalyzerPathResolver? resolver = null;
+            foreach (var current in AnalyzerPathResolvers)
             {
-                return _analyzerAssemblyInfoMap.ContainsKey(fullPath);
+                if (current.IsAnalyzerPathHandled(originalPath))
+                {
+                    resolver = current;
+                    resolvedPath = resolver.GetResolvedAnalyzerPath(originalPath);
+                    break;
+                }
             }
-        }
 
-        public void AddDependencyLocation(string fullPath)
-        {
-            CheckIfDisposed();
-
-            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
-            string simpleName = PathUtilities.GetFileName(fullPath, includeExtension: false);
+            var assemblyName = readAssemblyName(resolvedPath);
 
             lock (_guard)
             {
-                if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths))
+                if (_originalPathInfoMap.TryAdd(originalPath, (resolver, resolvedPath, assemblyName)))
```

Comment:
Can we assert that path in that won was the same one that we mapped here?

Reply:
Not sure what you're asking ... are you asking that we add something like 

```csharp
Debug.Assert(_originalPathInfoMap[orignialPath].ResolvedPath == resolvedPath);
```

That doesn't seem to add much value cause we're inside a `lock` so there is no chance that would fail unless the core invariant of "don't use _originalPathInfoMap` outside a `lock (_guard)`" was broken

Reply:
I don't see how the lock is relevant. There could have been two threads going: both calculated a resolved path outside of the lock, and now one of them is going to win. If they didn't calculate the same path, then something weird happened.

Reply:
> It can redirect the path 
>    /// originally passed to the compiler to a new path. Or it can take ownership of a path to prevent

There's no `or` here. Both of these scenarios take ownership of the path. Whether it does a no-op or not is irrelevant to the interface.

Reply:
> Can we assert that path in that won was the same one that we mapped here?
> I don't see how the lock is relevant 

The winning path is inside the lock, all other writes are in the lock, so inside the winning path in the lock there is no possibility of a loss. To assert there was the same we'd have to do so on the losing path or after we exit or reenter the lock. 

Are you saying create an `else` block that `Debug.Assert`?

Reply:
Yes, that is what I'm saying.